### PR TITLE
WIP: Fix getAllEntity lazy loading issue

### LINKED ISSUE #7463: Creating a Job with default JDL breaks entity access
##### **Overview of the issue**

When you try to create a Job entity the server will return a 500, and you won't be able to access that entity anymore.

##### **Motivation for or Use Case**

Bug with the default JDL.

##### **Reproduce the error**

Create a "vanilla" Angular project and import the default JDL provided by JDL Studio

##### **Related issues**

idk

##### **Suggest a Fix**

idk, I'll try to take a look this week.

##### **JHipster Version(s)**

```
test@0.0.0 /home/ippon/repos/tests/test
└── generator-jhipster@5.0.0-beta.0  -> /home/ippon/repos/generator-jhipster

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "com.mycompany.myapp",
      "nativeLanguage": "en"
    },
    "jhipsterVersion": "5.0.0-beta.0",
    "baseName": "test",
    "reactive": false,
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "cacheProvider": "ehcache",
    "enableHibernateCache": true,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "mysql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "maven",
    "enableSwaggerCodegen": false,
    "jwtSecretKey": "replaced-by-jhipster-info",
    "clientFramework": "angularX",
    "useSass": false,
    "clientPackageManager": "yarn",
    "applicationType": "monolith",
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": [
      "en"
    ]
  }
}
</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity Region (region) {
  regionName String
}
entity Country (country) {
  countryName String
}
/**
 * not an ignored comment
 */
entity Location (location) {
  streetAddress String,
  postalCode String,
  city String,
  stateProvince String
}
entity Department (department) {
  departmentName String required
}
/**
 * Task entity.
 * @author The JHipster team.
 */
entity Task (task) {
  title String,
  description String
}
/**
 * The Employee entity.
 */
entity Employee (employee) {
  /**
   * The firstname attribute.
   */
  firstName String,
  lastName String,
  email String,
  phoneNumber String,
  hireDate Instant,
  salary Long,
  commissionPct Long
}
entity Job (job) {
  jobTitle String,
  minSalary Long,
  maxSalary Long
}
entity JobHistory (job_history) {
  startDate Instant,
  endDate Instant,
  language Language
}

enum Language {
  FRENCH,
  ENGLISH,
  SPANISH
}

relationship OneToOne {
  Country{region} to Region,
  Location{country} to Country,
  Department{location} to Location,
  JobHistory{job} to Job,
  JobHistory{department} to Department,
  JobHistory{employee} to Employee
}
relationship OneToMany {
  /**
   * A relationship
   */
  Department{employee} to Employee{department},
  Employee{job} to Job{employee}
}
relationship ManyToOne {
  Employee{manager} to Employee
}
relationship ManyToMany {
  Job{task(title)} to Task{job}
}

dto Region, Country, Location, Department, Task, Employee, Job, JobHistory with mapstruct
service Region, Country, Location, Department, Task, JobHistory with serviceImpl
angularSuffix Region, Country, Location, Department, Task, Employee, Job, JobHistory with mySuffix
paginate Employee, JobHistory with infinite-scroll
paginate Job with pagination

</pre>
</details>

##### **Environment and Tools**

openjdk version "1.8.0_162"
OpenJDK Runtime Environment (build 1.8.0_162-8u162-b12-0ubuntu0.16.04.2-b12)
OpenJDK 64-Bit Server VM (build 25.162-b12, mixed mode)

git version 2.7.4

node: v8.11.1

npm: 5.6.0

yeoman: 2.0.1

yarn: 1.5.1

Docker version 1.13.1, build 092cba3

docker-compose version 1.8.0, build unknown

##### **Browsers and Operating System**

I don't think it matters.

### PR DESCRIPTION
**Pull request by agaspardcilia**
fix #7463 
This bug was caused by a lazy loaded collection that was nested inside another lazy loaded Object.
- Please make sure the below checklist is followed for Pull Requests.

- [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [x] Tests are added where necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 1]
```diff
@@ -58,6 +58,7 @@ import org.springframework.http.HttpStatus;
 import org.springframework.http.MediaType;
 <%_ } _%>
 import org.springframework.http.ResponseEntity;
+import org.springframework.transaction.annotation.Transactional;
```
Comment (ruddell) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs]: Transactional is only used for SQL databases, so add the EJS condition `<%_ if (databaseType === 'sql') { _%>`

File: generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs [hunk 2]
```diff
@@ -148,7 +149,8 @@ public class <%= entityClass %>Resource {
      * @return the ResponseEntity with status 200 (OK) and the list of <%= entityInstancePlural %> in body
      */
     @GetMapping("/<%= entityApiUrl %>")
-    @Timed<%- include('../../common/get_all_template', {viaService: viaService}); -%>
+    @Timed
+    @Transactional(readOnly = true)<%- include('../../common/get_all_template', {viaService: viaService}); -%>
```
Comment (ruddell) [inline generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs]: Same comment as above,
```
    <%_ if (databaseType === 'sql') { _%>
    @Transactional(readOnly = true)
    <%_ } _%>
```

**Comment by agaspardcilia** [comment]
@ruddell Thanks for the quick review, it's the first time that I'm touching anything in the back-end. :relaxed:

**Comment by deepu105** [comment]
I think we deliberately removed Transactional from Resources. May be we should just enable service layer if DTO is selected and vice versa

**Comment by jdubois** [comment]
Yes the Web layer should not be transactional

**Comment by jdubois** [comment]
I'm removing this because the Web layer shouldn't be transactional, so that's not a good solution - @agaspardcilia I'll be at the office on Wednesday/Thrusday/Friday could you show me the issue? So we can find a better solution

**Comment by agaspardcilia** [comment]
@jdubois If forgot to close this PR indeed. This problem is due to a permitted option that is about to get removed from JDL. The both of us prolly don't need to do anything about it.

**Pull request closed without merging**